src/core/ruby_version.rs

```diff
@@ -1289,71 +1516,126 @@ macro_rules! get_cfunc_name_unsupported(
+            if method_name.is_empty() {
+                return method_name.to_string();
+            }
 
-            // if VM_FRAME_TYPE($cfp->flag) != VM_FRAME_MAGIC_CFUNC
-            if frame_flag & 0xffff0001 != 0x55550001 {
-                return Err(format_err!("Not a C function control frame").into());
+            if !class_path.is_empty() {
+                let join_char = if singleton { "." } else { "#" };
+                return format!("{}{}{}", class_path, join_char, method_name);
             }
 
+            method_name.to_string()
+        }
+
+        fn locate_method_entry<T>(
+            ep: &*const usize,
+            source: &T,
+        ) -> Result<usize> where T: ProcessMemory {
+            const VM_ENV_FLAG_LOCAL: usize = 0x2;
+            let mut ep = ep.clone() as *mut usize;
+            let mut env_flags: usize = source.copy_struct(ep as usize).context(ep as usize)?;
             let mut env_specval: usize = unsafe {
                 source.copy_struct(ep.offset(-1) as usize).context(ep.offset(-1) as usize)?
             };
             let mut env_me_cref: usize = unsafe {
-                source.copy_struct(ep.offset(-2) as usize).context(ep.offset(-1) as usize)?
+                source.copy_struct(ep.offset(-2) as usize).context(ep.offset(-2) as usize)?
             };
 
-            // #define VM_ENV_FLAG_LOCAL 0x02
-            while env_specval & 0x02 != 0 {
-                if !check_method_entry(env_me_cref, source)?.is_null() {
-                    break;
+            if env_flags & VM_ENV_FLAG_LOCAL == 0 {
+                //println!("ALREADY LOCAL {:X}", env_me_cref);
+            }
+            while env_flags & VM_ENV_FLAG_LOCAL != 0 {
+                //println!("CHECK NON LOCAL");
+                let me = check_method_entry(env_me_cref, false, source)?;
+                if me != 0{
+                    return Ok(me);
                 }
                 unsafe {
-                    ep = ep.offset(0) as *mut usize;
+                    // https://github.com/ruby/ruby/blob/v3_4_5/vm_core.h#L1356
+                    // we must strip off the GC marking bits from the EP, and mimic
+                    // https://github.com/ruby/ruby/blob/v3_4_5/vm_core.h#L1501
+                    ep = (env_specval.clone() & !0x03) as *mut usize;
+                    env_flags = source.copy_struct(ep as usize).context(ep as usize)?;
                     env_specval = source.copy_struct(ep.offset(-1) as usize).context(ep.offset(-1) as usize)?;
                     env_me_cref = source.copy_struct(ep.offset(-2) as usize).context(ep.offset(-2) as usize)?;
                 }
             }
 
-            let imemo: rb_method_entry_struct = source.copy_struct(env_me_cref).context(env_me_cref)?;
+            let me = check_method_entry(env_me_cref, true, source)?;
+            //println!("ME {:X} {:X}", me, env_me_cref);
+            Ok(me)
+        }
+
+        fn check_method_entry<T: ProcessMemory>(
```

Comment:
This function was also pretty broken. I rewrote it to mimic the logic in ruby's vm_insnhelper.c